### ISSUE TITLE
High CPU usage and possible memory leak on docker logs with SpringBoot applications.

### ISSUE THREAD

**Issue by pythianarceri** [docker/docker#30995]
**Description**
High CPU usage and possible memory leak on docker logs with SpringBoot applications.
We have a rancher environment with several containers running SpringBoot Applications, we have an ELK stack to which we send container logs using logspout with the logstash adapter.
For some unknown reason, if we allow logspout to collect logs from these containers dockerd cpu utilization and memory utilization start to climb (i've seen load averages of up to 70), to the point that the server needs to be rebooted in order to recover it.
Furthermore when looking at the container logs using "docker logs"  i get different output every time i run the command, and that output last ten lines never match what docker logs --tail=10 is displaying
```[rancher@ip-10-30-0-193 ~]$ docker logs c61f0d662ef5

[rancher@ip-10-30-0-193 ~]$ docker logs c61f0d662ef5

[rancher@ip-10-30-0-193 ~]$ docker logs c61f0d662ef5

[rancher@ip-10-30-0-193 ~]$ docker logs c61f0d662ef5

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
[rancher@ip-10-30-0-193 ~]$ docker logs c61f0d662ef5

[rancher@ip-10-30-0-193 ~]$ docker logs c61f0d662ef5

  .   ____          _            __ _ _
[rancher@ip-10-30-0-193 ~]$ docker logs --tail=10 c61f0d662ef5
2017-02-10 17:50:55.161  INFO 1 --- [           main] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/**] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2017-02-10 17:50:56.193  INFO 1 --- [           main] o.s.w.s.handler.SimpleUrlHandlerMapping  : Mapped URL path [/**/favicon.ico] onto handler of type [class org.springframework.web.servlet.resource.ResourceHttpRequestHandler]
2017-02-10 17:51:01.775  INFO 1 --- [           main] o.s.j.e.a.AnnotationMBeanExporter        : Registering beans for JMX exposure on startup
2017-02-10 17:51:01.921  INFO 1 --- [           main] o.s.c.support.DefaultLifecycleProcessor  : Starting beans in phase 0
2017-02-10 17:51:19.895  INFO 1 --- [           main] /                                        : Initializing Spring FrameworkServlet 'dispatcherServlet'
2017-02-10 17:51:19.917  INFO 1 --- [           main] o.s.web.servlet.DispatcherServlet        : FrameworkServlet 'dispatcherServlet': initialization started
2017-02-10 17:51:20.495  INFO 1 --- [           main] o.s.web.servlet.DispatcherServlet        : FrameworkServlet 'dispatcherServlet': initialization completed in 570 ms
2017-02-10 17:51:21.140  INFO 1 --- [           main] o.eclipse.jetty.server.ServerConnector   : Started ServerConnector@48433ada{HTTP/1.1}{0.0.0.0:8100}
2017-02-10 17:51:21.217  INFO 1 --- [           main] .s.b.c.e.j.JettyEmbeddedServletContainer : Jetty started on port(s) 8100 (http/1.1)
2017-02-10 17:51:21.272  INFO 1 --- [           main] com.bps.statement.analysis.Application   : Started Application in 161.074 seconds (JVM running for 178.074)
```
Is it possible that multiline messages are somehow "confusing" docker ?

**Steps to reproduce the issue:**
1. Deploy springboot application
2. Configure logspout to collect logs from the application and send them to logstash
3. Server performance starts degrading

**Describe the results you received:**
Server will ultimately crash, dockerd can be seen using 80%+ CPU and its memory utilization starts climbing.

**Describe the results you expected:**
- dockerd should not hog the cpu/memory
- `docker logs` output should be the same everytime the command is executed

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**
```
[rancher@ip-10-30-0-193 ~]$ docker version
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 23:26:11 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 23:26:11 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
[rancher@ip-10-30-0-193 ~]$ docker info
Containers: 62
 Running: 40
 Paused: 0
 Stopped: 22
Images: 94
Server Version: 1.12.3
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local rancher-nfs
 Network: host null bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.24-rancher
Operating System: RancherOS v0.7.1
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 14.94 GiB
Name: ip-10-30-0-193.ec2.internal
ID: DVSV:NLW5:VD5U:UOYY:FIOB:4RTH:7NGQ:GIBO:KKTR:VPYO:J6A2:4SZE
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
AWS

**Comment by AkihiroSuda** [comment]
Can you please provide images to hit this issue?

**Comment by pythianarceri** [comment]
Let me see what can i do as the images may contain confidential information. I need to run them by our customer.

**Comment by cpuguy83** [comment]
A stack dump would also be incredibly helpful.
You can get this by sending `SIGUSR1` to the dockerd process. Trace will be in the daemon logs.

**Comment by thaJeztah** [comment]
@cpuguy83 I think they're written to separate files now (but the logs show where the files can be found)

**Comment by pythianarceri** [comment]
Attaching dockerd stack trace as requested, let me know what you find
[dockerd_stacktrace.txt](https://github.com/docker/docker/files/774190/dockerd_stacktrace.txt)

**Comment by cpuguy83** [comment]
@pythianarceri Is this from a dockerd that is using excessive mem/cpu?

**Comment by pythianarceri** [comment]
Yes, i enabled lospout and started the "offending" container, when i took the stack trace dockerd was using 70% CPU

**Comment by cpuguy83** [comment]
@pythianarceri Do you have some details on the offending container (is it a particular one?)?
Can you provide the inspect output?

**Comment by cpuguy83** [comment]
In the provided trace I only see a bunch of blocked goroutines (blocked on io, syscall, select, or chan receive). Nothing obviously taking CPU that I can find.

**Comment by pythianarceri** [comment]
I think the customer has made some modifications to the container, which apparently is not generating any logs now, i'll keep you updated.

**Comment by pythianarceri** [comment]
Customer has performed some modification to their app code that have affected logging, i'm unable to reproduce the issue anymore (but i'm not seeing much logs either), i'm trying to revert to the old app code to try to reproduce the issue.

**Comment by cpuguy83** [comment]
@pythianarceri Thanks for continuing to look into this.

**Comment by cpuguy83** [comment]
We identified an issue in 1.13 causing a CPU spin with `docker logs` in the daemon.
This will be fixed in the next patch release.

Fixed by #31070

Closing but if you run into another issue after applying that patch please let us know.

Thanks again!

**Comment by Dean-Christian-Armada** [comment]
I'm using 180.05 Docker version but it's occurring in my Swarm everytime I use logspout

**Comment by cpuguy83** [comment]
@Dean-Christian-Armada Can you open a new issue with repo steps? Almost certainly this is a different problem.

Thanks!
